### DISCUSSION TITLE
How can I transform metrics received through another channel to OTL metrics?

### DISCUSSION THREAD

**Discussion by unwork-ag** [open-telemetry/opentelemetry-dotnet#3466] [Q&A]
A couple of years back we have built a much simplified home-grown kind of collector and now want to integrate that with OpenTelemetry. Our collector receives json data through REST calls and one type of data is performance information. We would like to send this to an OTel collector using OTLP as metric data (i.e. histogram) but we are not sure how to feed the data that we receive into the OTel pipeline. 

Things we have considered so far:

- Creating a meter in our own collector and adding histograms for each type of performance record we get. This makes the record look like it was created in the receiving service (our custom collector) which is wrong. We have all the original data available - but we didn't find a way to put this into the metrics data.
- Building a custom metrics reader that takes our data and transforms it into OpenTelemetry metrics. This looked promising but we unfortunately we cannot construct the Metrics type that we need to pass to the exporter (ctor is internal: https://github.com/open-telemetry/opentelemetry-dotnet/blob/23609730ddd73c860553de847e67c9b2226cff94/src/OpenTelemetry/Metrics/Metric.cs) 

I think we are not approaching this correctly. Does someone have a suggestion how to do this the right way?

**Comment by Qinyouzeng** [comment]
We also need a similar solution now. The difference is that after reading opentelemrtry’s indicator collection, I found that if you want to add and export custom indicators to otlp, there are currently two methods:
1. Use a unique `Meter` to manage all your indicators, and then when using Opentelemetry Metric, register the meter Name, [Meter] to create indicators (https://github.com/open-telemetry/opentelemetry-dotnet/ blob/main/docs/metrics/customizing-the-sdk/Program.cs),
Register `Meter`
``` C#
services.AddOpenTelemetry().WithMetrics(builder =>
         {
             builder.AddMeter("your meter name1","your meter name2");

             builder. AddRuntimeInstrumentation()
             .AddAspNetCoreInstrumentation();
             .AddHttpClientInstrumentation();
         }).StartWithHost();
         return services;
```
2. [Reference](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInMetricsListener.cs), replaced by `AddMeter` with `AddInstrumentation`
``` C#
       var myMtrics = new MyMtrics();
       builder.AddInstrumentation(myMtrics);
```
Finally, export to `OTLP` uniformly.
Mainly, the above-mentioned `AddMeter` and `AddInstrumentation` documents have relatively few descriptions. You need to look at the source code indicator collection. I also hope that the development team’s documents will be more detailed.

**Reply by unwork-ag** [reply]
Hello @Qinyouzeng - thanks for taking the time to reply to my question. I have to admit I'm not sure I understand what you are suggesting. The particular problem I have is that I want to make sure that information like when and where a metrics data point was collected is preserved. If I just create new measurements, they appear as if they were recorded in my service instead of in the original caller (and at another time).

Your suggestion 1 looks like the standard way to capture custom metrics from within a process. This would lead to the problem that measurements get misleading information about the origin of the measurement.

I don't completely get where you are heading to with suggestion 2: `AddInstrumentation` usually takes an instrumentation factory. What is your `MyMtrics` class and how does it feed measurements into the pipeline?

**Comment by Qinyouzeng** [comment]
I misunderstood your needs. I don't know if your needs are met, but they are met according to the original:
You can use `Meter`
``` C#
        //define a meter  refrence https://learn.microsoft.com/en-us/dotnet/core/diagnostics/metrics-collection
        Meter meter = new("MyCompany.MyProduct.MyLibrary", "1.0");
        //create custom metrics
        var metricRequest = meter.CreateHistogram<int>("my_custom_metric_request", "", "metrics description");
        var metricResponse = meter.CreateHistogram<double>("my_custom_metric_response", "", "metrics description");

        // config AspNetCoreInstrumentationOptions        
        services.AddOpenTelemetry().WithTracing(builder =>
        {
            builder.SetSampler(new AlwaysOnSampler());
            builder.AddAspNetCoreInstrumentation(options =>
            {
                //http request
                options.EnrichWithHttpRequest = (Activity activity, HttpRequest httpRequest) =>
                {
                    int value = 1;
                    var tags = new List<KeyValuePair<string, object?>>();
                    //your any tags can add here,from httpRequest or fixed
                    tags.Add(KeyValuePair.Create("my_custom_time", (object?)DateTime.Now));
                    tags.Add(KeyValuePair.Create("my_custom_position", (object?)"Hangzhou"));
                    metricRequest.Record(value, tags.ToArray());
                };

                options.EnrichWithHttpResponse = (Activity activity, HttpResponse httpResponse) =>
                {
                    double value = activity.Duration.TotalMilliseconds;
                    var tags = new List<KeyValuePair<string, object?>>();
                    tags.Add(KeyValuePair.Create("my_custom_time", (object?)DateTime.Now));
                    tags.Add(KeyValuePair.Create("my_custom_position", (object?)"Hangzhou"));
                    metricResponse.Record(value, tags.ToArray());
                };
            }).AddOtlpExporter();
        });

        services.AddOpenTelemetry().WithMetrics(builder => {
            //add my custom metrics to otlp exporter
            builder.AddMeter(meter.Name)
            .AddAspNetCoreInstrumentation()
            .AddOtlpExporter();            
        });
```
hope it can help you.
